File: leetcode/linklist/Copy List with Random Pointer.go
package linklist


/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

//对于此题，或者说对于多达数linklist的题，1.就是检查是否head为nil
//2,时刻警惕着当前指针指向nil
func copyRandomList(head *Node) *Node {
	if head == nil{
		return head
	}
	cur := head

	for;cur != nil;{
		tmp := cur.Next
		cur.Next = &Node{cur.Val, nil , nil}
		cur.Next.Next = tmp
		cur = tmp
	}

	cur = head
	for;cur != nil;{
		if cur.Random != nil{
			cur.Next.Random = cur.Random.Next
		}
		cur = cur.Next.Next
	}

	cur = head
	copyhead := head.Next
	copycur := copyhead

	for;cur != nil;{
		cur.Next = copycur.Next
		cur = copycur.Next
		if cur != nil{
			copycur.Next = cur.Next
			copycur = cur.Next
		}
	}

	return copyhead
}

File: leetcode/archived/C++/Copy_list_with_random_pointer.cpp
/*Defintion for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
	public:
		RandomListNode *copyRandomList(RandomListNode *head) {

			RandomListNode *res = NULL, *cur = head;
			if(!cur)
				return NULL;

			//insert one node betweent two nodes
			while(cur)
			{
				RandomListNode *temp = cur->next;
				cur->next = new RandomListNode (cur->label);
				cur->next->next = temp;
				cur = temp;
			}

			cur = head;

			//copy random pointer
			while(cur)
			{
				if((cur->random))
					cur->next->random = cur->random->next;
				else
					cur->next->random = NULL;
				cur = cur->next->next;
			}

			cur = head;
			res = cur->next;

			//remove duplicate ndoes
			while(cur)
			{
				RandomListNode *cur1 = cur->next;
				cur->next = cur->next->next;
				cur = cur->next;
				if(cur)
					cur1->next = cur->next;
				else
					cur1->next = NULL;
			}

			return res;
		}
};
